temporalio/converter.py

```diff
@@ -0,0 +1,362 @@
+"""Base converter and default implementations for conversion to/from values/payloads."""
+
+import dataclasses
+import inspect
+import json
+from abc import ABC, abstractmethod
+from typing import Any, Optional, Tuple, Type
+
+import google.protobuf.json_format
+import google.protobuf.message
+import google.protobuf.symbol_database
+
+import temporalio.api.common.v1
+
+
+class DataConverter(ABC):
+    """Base converter to/from multiple payloads/values."""
+
+    @abstractmethod
+    async def encode(self, values: list[Any]) -> list[temporalio.api.common.v1.Payload]:
+        """Encode values into payloads.
+
+        Args:
+            values: Values to be converted.
+
+        Returns:
+            Converted payloads. Note, this does not have to be the same number
+            as values given, but at least one must be present.
+
+        Raises:
+            Exception: Any issue during conversion.
```

Comment:
Can you provide your reasoning?

I'd rather use https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings. It's cleaner to read, less boilerplate, supported by Sphinx like the others, etc. Having looked at them all and tooling support for them all, I think this approach offers the most benefits.

Reply:
The reST format is the most common in my experience, and is quite a bit more compact than this style. FWIW it's recommended by this PEP https://www.python.org/dev/peps/pep-0287/

Reply:
That PEP does say reST which we of course use but doesn't say how to do parameters and the like. Doc strings are free-form. I've been looking at modern popular projects and I definitely see a mix. Projects like mypy, black, and other newer ones tend to use the Google style (or [similar](https://pythonhosted.org/an_example_pypi_project/sphinx.html#full-code-example)) and older ones like flask tend to use Sphinx-based approaches. Of course PyCharm supports the Google format since it is so prevalent in modern code. I personally find the Google ones ([example](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) easier to read, but there are also the Numpy ones ([example](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html)). What @bergundy linked is just an example of how to use "Epytext" format, not a suggestion.

Even the [Sphinx docs](https://www.sphinx-doc.org/en/master/tutorial/describing-code.html) are a bit underspecified. E.g. those docs say `:raise <err>` yet https://stackoverflow.com/questions/34160968/python-docstring-raise-vs-raises. Basically, like the rest of Python, it's a big mess of underdocumented stuff.

I would prefer the Google approach, but if the Sphinx or Numpy approach is preferred we can.

Reply:
My rationale is that's what pycharm defaults to and from my experience is the most popular format.
I don't have a strong preference, my request was based on personal experience.

Reply:
> that's what pycharm defaults

I think it defaults to supporting all of them. That example you linked works _after_ you had chosen `Epytext` and it even shows how it might look if you had selected `reStructuredText`.

> from my experience is the most popular format.

This hasn't been my experience surveying the landscape, especially Python 3. In fact, I've found of the 4 primary formats: `reStructuredText`, `Epytext`, `NumPy`, and `Google` (that of course PyCharm supports), Epytext is the least popular.

Reply:
Surveying the landscape, this format is quite popular in new projects. We can change if we decide to, but there seems to be no compelling reason to do so.